h/test/app_test.py

```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+
+import mock
+import pytest
+
+from pyramid.testing import DummyRequest
+
+from h import app
+
+class TestPublishAfterCommitEvents(object):
+    def test_skips_publishing_events_on_exception(self):
+        request = self.makerequest(exception=ValueError('exploded!'))
+        app.publish_after_commit_events(request, None)
+        assert not request.after_commit_events.publish_all.called
+
+    def test_skips_publishing_events_when_no_events_queued(self):
+        request = self.makerequest(events=[])
+        app.publish_after_commit_events(request, None)
+        assert not request.after_commit_events.publish_all.called
+
+    def test_publishes_events(self):
+        request = self.makerequest(events=[mock.Mock()])
+        app.publish_after_commit_events(request, None)
+        assert request.after_commit_events.publish_all.called
+
+    def makerequest(self, exception=None, events=None):
+        event_queue = mock.MagicMock(spec=app.EventQueue)
+        if events:
+            event_queue.__len__.return_value = len(events)
+
+        request = DummyRequest(tm=mock.MagicMock(),
+                               exception=exception,
+                               after_commit_events=event_queue)
+        return request
+
+class TestEventQueue(object):
+    def test_publish_all_pops_from_left(self, popleft):
```

Comment:
I wonder if what we really want to test here is that `self.request.registry.notify()` is called with the events in the right order? And/or that there are no more events on the queue after `publish_all()` has completed successfully? Testing that it calls `popleft()` seems a little like testing an internal implementation details of the method itself to me, and I'm not sure what the purpose of the test it.